protos/core.proto

```diff
@@ -247,6 +247,190 @@ message NotebookRuntimeOptions {
   }
 }
 
+// Data Preparation Related entries
+message DataPreparation {
+  // Data preparatiohs can have more than 1 output
+  repeated Target targets = 1;
+
+  repeated Target canonical_targets = 2;
+
+  repeated string tags = 3;
+
+  repeated Target dependency_targets = 4;
+
+  string file_name = 5;
+
+  bool disabled = 6;
+
+  DataPreparationDefinition data_preparation = 7;
+}
+
+message DataPreparationDefinition {
+  repeated DataPreparationNode nodes = 1;
+  DataPreparationGenerated generated = 2;
+}
+
+message DataPreparationNode {
+  string id = 1;
+  DataPreparationNodeSource source = 2;
+  repeated DataPreparationNodeStep steps = 3;
+  DataPreparationNodeGenerated generated = 4;
+  // Destination BigQuery table(s) are defined within the data preparation
+  optional DataPreparationNodeDestination destination = 5;
+}
+
+message DataPreparationNodeSource {
+  oneof source {
+    string node_id = 1;
+    TableReference table = 2;
+    Join join = 3;
+  }
+}
+
+message DataPreparationNodeDestination {
+  oneof destination {
+    TableReference table = 1;
+  }
+}
+
+message DataPreparationNodeStep {
+  string id = 1;
+  string description = 2;
+  oneof definition {
+    ColumnStep column_step = 3;
+    FilterStep filter_step = 4;
+  }
+  DataPreparationNodeStepGenerated generated = 5;
+}
+
+message ColumnStep {
+  string column_name = 1;
+  Expression expression = 2;
+}
+
+message FilterStep {
+  Expression expression = 1;
+}
+
+message Expression {
+  oneof expression {
+    string sql = 1;
+  }
+}
+
+message Join {
+  string left_node_id = 1;
+  string right_node_id = 2;
+  JoinType join_type = 3;
+  JoinCondition join_condition = 4;
+}
+
+enum JoinType {
+  JOIN_TYPE_UNSPECIFIED = 0;
+  JOIN_TYPE_INNER = 1;
+  JOIN_TYPE_FULL_OUTER = 2;
+  JOIN_TYPE_LEFT = 3;
+  JOIN_TYPE_RIGHT = 4;
+}
+
+message JoinCondition {
+  oneof condition {
+    Expression expression = 1;
+    JoinKeys keys = 2;
+  }
+}
+
+message JoinKeys {
+  repeated JoinKey keys = 1;
+}
+
+message JoinKey {
+  string left_column = 1;
+  string right_column = 2;
+}
+
+message TableReference {
+  string project = 1;
+  string dataset = 2;
+  string table = 3;
+}
```

Comment:
It seems strange to duplicate the table target like this. Target is already defined https://github.com/dataform-co/dataform/blob/97123faf432781480bee6f888a4881d934aff349/protos/core.proto#L47

Reply:
It looks similar, it just uses BigQuery specific terminology. Data Preparation is based on a BigQuery Source/Destination, thus we defined the protos with BQ specific terminology.